Unambiguous uncommitted files CLI ID and cleanup

jonathantanmy2 opened the pull request:
Next one in the series. The next step for me is either

* look at `but-hunk-assignment`, but it itself uses `but-hunk-dependency`, so investigation on that might need to wait until we figure out how to move forward with #11368 , or
* treat the output of the hunk assignments calculation as a black box and refactor `IdMap` to only require `RefInfo` and the hunk assignments

My inclination is to do the latter. Something needs to know how to generate the `RefInfo` and the hunk assignments it needs, though - unfortunately I think that will have to be `Context`.

As for this commit chain itself, note that there is an error message change in "id: unambiguous uncommitted file CLI IDs", as described in its commit message. Take a look and see if that's reasonable. Also, how do you want it tested? I saw `crates/but/src/legacy/id/tests.rs` but I don't know how to add uncommitted files (both in the `00` area and in a stack) without using `but` commands (which would require an integration test). Or, we can wait until I've refactored `IdMap` to take in the hunk assignments directly, and only test this part then.

jonathantanmy2 committed:
id: unambiguous uncommitted file CLI IDs

They also do not collide with committed file IDs or branch IDs.

Sample output:

```
╭┄000 [Unassigned Changes]
┊   g0 A d
┊
┊╭┄ux [b00] (no commits)
├╯
┊
┊╭┄wo [work]
┊│   h0 A e
┊●   89f03ba test commit
┊│     j0 A a
┊│     k0 A b
┊│     l0 A c
├╯
┊
┊╭┄i0 [i0] (no commits)
├╯
┊
┴ dba337d (common base) [origin/main] 2025-11-11 Merge pull request #2254 from ralphmodales/credent
```

I had to do some contortions in `context_info()` to ensure that
`repo`, and hence the indirect borrowing of `ctx`, was dropped
before `but_hunk_assignment::assignments_with_fallback()`, which
needs to borrow `ctx` mutably, was called. This will be fixed once
`context_info()` does not take `ctx`, but only what it needs (most
likely, just the `RefInfo` and the hunk assignments).

There is also a change to an error message in a test
(`crates/but/tests/but/journey.rs`), likely due to uncommitted files
being read at `IdMap` construction time, instead of when it is used to
parse a string.

TODO need to write tests

jonathantanmy2 committed:
id: rename IdDb to IdMap, needs mut Context

In preparation for a future commit which needs a mut Context to
construct the IdDb, take a mut Context instead of a non-mut one.

While we're at it, since many lines will need to be changed anyway,
rename IdDb to a more natural name, IdMap.

jonathantanmy2 committed:
id: cleanup `parse_str` branch handling

In `parse_str`, despite the comment near `self.find_branches_by_name`,
`self.find_branches_by_name` actually handles partial branch name lookup
already, which means that the other branch searches are redundant.
Remove them, which also allows a lot of other code cleanup.

In particular, `Context` is no longer needed when parsing strings - only
when generating the `IdMap`. This makes it clear that the only inputs
are the `RefInfo` and the hunk assignments.

Byron commented on crates/but/src/legacy/id/mod.rs:
```diff
@@ -18,56 +17,95 @@ struct ContextInfo {
+    fn borrow(&self) -> &str {
+        &self.id
+    }
+}
```
I really like these `Borrow` implementations by the way and I wasn't aware of how this can be used to use sets a little like maps!
It's great.

CC @krlvi

Byron commented on crates/gitbutler-branch-actions/src/base.rs:
```diff
@@ -74,12 +74,12 @@ fn go_back_to_integration(ctx: &Context, default_target: &Target) -> Result<Base
         but_core::worktree::safe_checkout(
             gix_repo.head_id()?.detach(),
-            tree_to_checkout_to_avoid_ref_update.detach(),
+            tree_to_checkout_to_avoid_map_update.detach(),
```
I think this was accidentally renamed, good point.

jonathantanmy2 replied:
Ah...find-and-replace error. Fixed.

Byron committed:
refactor

- make clear that `ContextInfo` is an intermediary, a utility
  by immediately destructuring it.
- address copilot comments

Byron commented on crates/but/tests/but/journey.rs:
```diff
@@ -106,7 +106,10 @@ Caused by:
+Error: errors.projects.default_target.not_found
+
+Caused by:
+    there is no default target
```
You probably mentioned it, but this feels like a regression.
However, it was also coincidentally 'good' before, and I think this can be ignored until single-branch mode is properly supported.
Now the ID map seems to trigger another related error early, somewhere in old code probably.

jonathantanmy2 commented on crates/but/src/legacy/id/mod.rs:
```diff
@@ -230,13 +275,8 @@ impl IdDb {
                     id: s.to_string(),
                 });
             }
```
Fixed.

Byron reviewed (approved):
Thanks again, and I agree that making `IdMap` solid is something that can be done while `but-hunk-dependency` is pending.

I love where this is going, and thought it was interesting that the `&mut Context` requirement immediately signals database access (i.e. assignments) just because all DB access needs mutability right now. It's a major issue, and I want to address it till Monday, to let read-only remain read-only.

Regarding testing, there is `but_hunk_assignment::assign(ctx, reqs, None)?` to make assignments, and unit-tests can be made to assign after the initial fixture was created. It's writable, after all, and it should work.

Let me try to add an example test before merging.

Byron committed:
IdMap unit'y test with assignments (and add `but-testsupport::Sandbox`)

This also makes the Sandbox more generally available, in support of a more
generalised testing style of somewhat higher-level functions that use the context.

Byron commented:
That was a quite a refactor, which ultimately led to a generalised `Sandbox` implementation which should be useful even in plumbing crates. It's all about isolation and the kinds of components it can produce, and in `but-api` mode, opted in by feature toggle, it will also isolate application data which right now is really only needed for `projects.json` (and of course, for the clients under test).

Right now, there are no non-but-api users of the Sandbox, but I think it's good to have the feature toggle to be forced to keep the 'plumbing-vs-api' distinction alive in code as well, and clearly separated.

Byron commented on crates/but-meta/src/legacy.rs:
```diff
@@ -242,16 +269,19 @@ impl Snapshot {
         }
 
-        let stack_ids_to_put_in_workspace: Vec<_> = sideeffect_free_meta
+        let stack_ids_to_remove_from_workspace: Vec<_> = sideeffect_free_meta
```
Falling in love here!

Byron committed:
Address Copilot review notes

Byron commented on crates/but-testsupport/Cargo.toml:
```diff
@@ -7,21 +7,37 @@ publish = false
 rust-version = "1.89"
 
 [lib]
```
Fair - it can be confusing to write a unit test and not see it running, let's make this more prominent.
